### ISSUE TITLE
Persistent queue can have data loss if last send was in error state before restarting/shutdown

### ISSUE THREAD

**Issue by splunkericl** [open-telemetry/opentelemetry-collector#6771]
**Describe the bug**
Recently, a [new fix](https://github.com/open-telemetry/opentelemetry-collector/issues/6692) was added to prevent duplicate data by allowing writing to storage during shutdown. However, this has caused a side effect of data loss if the last send in the queue was in an error state. As `OnProcessingFinished` is always executed and the item is always removed from dispatched list even if the send was an error.

Examples of shutdown that can cause the problem:
- shut down would cause queue to stop looping processing of the item: https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queued_retry.go#L438
- the error will propagate to the callback function but ignored: https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queued_retry.go#L205
- OnProcessingFinished will be triggered:
  - persistent queue callback function: https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/internal/persistent_storage.go#L272
  - execution will remove the item from `currentlyDispatchedItems`
- After restarting, the item won't be processed again since it was considered dispatched 

**Steps to reproduce**
1. Create a pipeline with an exporter that is always failing to export
2. send a log entry to the pipeline
3. restart the pipeline with a new settings with an exporter that is able to send data

**What did you expect to see?**
After step 3 restart, it is expected the log is sent to the destination

**What did you see instead?**
The item was not seen in the exporter destination again. 

**What version did you use?**
Version: v0.67.0

**What config did you use?**
N/A

**Environment**
OS: mac
Compiler(if manually compiled): go 1.18

**Additional context**
A suggestion to fix the problem:
- Update `OnProcessingFinished` parameter to contain `error`
- pass in error from `consumerSender.send` to `OnProcessingFinished`
- Update persistent queue to not update storage if error is returned.

```
	qrs.queue.StartConsumers(qrs.cfg.NumConsumers, func(item internal.Request) {
		err := qrs.consumerSender.send(item)
		item.OnProcessingFinished(err)
	})
```

**Comment by bogdandrutu** [comment]
cc @dmitryax since you approved that PR you own it :)

**Comment by splunkericl** [comment]
I will be happy to make the changes as well if you guys don't have bandwidth :)

**Comment by dmitryax** [comment]
@splunkericl Of course, please feel free to take it

**Comment by atoulme** [comment]
Marking as fixed with #6815 - please reopen if more work is needed
